File: src/svmtest.cpp
#include <vector>
#include "SVM.h"

#define BUF_LEN 256

using namespace QuadProgPP;

// Vectorを標準出力する
template<typename T>
void printVector(const Vector<T>& v) {
  for (int i = 0; i < v.size(); ++i) {
    std::cout << v[i] << " ";
  }
  std::cout << std::endl;
}

// vectorをVectorに変換
template<typename T>
Vector<T> extendVector(const std::vector<T>& v){
  Vector<T> V;
  V.resize(v.size());
  for (size_t i = 0; i < v.size(); i++)
    V[i] = v[i];
  return V;
}

// カーネル名からKernelオブジェクト生成
Kernel *kernelWithName(const char *name){
  if (strcmp(name, "DotProd") == 0){
    return new DotProd();
  } else if (strcmp(name, "Gaussian") == 0){
    return new Gaussian(10);
  } else if (strcmp(name, "Polynomial") == 0){
    return new Polynomial(2);
  } else if (strcmp(name, "Sigmoid") == 0){
    return new Sigmoid(0.01, -3);
  } else {
    throw 1;
  }
}

int main(int argc, char *argv[]) {
  std::vector<std::vector<double> > x;
  std::vector<double> *x_row;
  std::vector<double> y;

  // コマンドライン引数のカーネル名kernel_nameから、Kernelを生成
  Kernel *kernel;
  const char *kernel_name;  
  try {
    if(argc == 1) {
      kernel_name = "DotProd";
    } else if(argc == 3){
      char *p;
      p = argv[1];
      if(*p == '-'){
        p++;
        switch(*p){
          case 'k':
            kernel_name = argv[2];
            break;
          default:
            throw 0;
        }
      } else {
        throw 0;
      }
    } else {
      throw 0;
    }
    
    kernel = kernelWithName(kernel_name);

  } catch(int n){
    switch(n){
      case 0:
        std::cout << "Invalid arguments" << std::endl;
        break;
      case 1:
        std::cout << "Invalid kernel name" << std::endl;
        break;
    }
    return 1;
  }

  // サンプルデータを標準入力から受け、vectorに格納
  char buf[BUF_LEN];
  char *tp;
  while(fgets(buf, BUF_LEN, stdin) != NULL) {
    x_row = new std::vector<double>();

    tp = strtok(buf, " ");
    x_row->push_back(atof(tp));

    while(tp != NULL) {
      tp = strtok(NULL, " ");
      if (tp != NULL) {
        x_row->push_back(atof(tp));
      } else {
        y.push_back(x_row->back());
        x_row->pop_back();
      }
    }

    x.push_back(*x_row);
  }

  // サンプルデータ点をXに格納
  Matrix<double> X;
  X.resize(x.size(), x[0].size());
  for (size_t i = 0; i < x.size(); i++)
    for (size_t j = 0; j < x[0].size(); j++)
      X[i][j] = x[i][j];
  // サンプルデータ点に対するクラスをYに格納
  Vector<double> Y = extendVector(y);

  // SVMオブジェクトを生成 コンストラクタ内で学習をする
  SVM svm(X, Y, kernel);

  // パラメータを出力
  svm.printAlpha();
  svm.printTheta();

  /*
  std::vector<double> r;
  for (int i = 0; i < X.extractColumn(0).size(); ++i) {
    r.push_back(svm.discriminate(X.extractRow(i)));
  }
  Vector<double> R = extendVector(r);

  std::cout << "result:" << std::fixed << std::setprecision(0) << std::endl;
  printVector(Y);
  */

  return 0;
}

File: src/svmtest.py
#coding:utf-8
import sys
import numpy as np
from scipy.linalg import norm
from SVM import SVM

# カーネルトリック
def gaussianKernel(x, y, sigma=10):
  return np.exp(-norm(x-y)**2 / (2*sigma**2))
def sigmoidKernel(x, y, a=0.002, b=-1.5):
  return np.tanh(a * np.dot(x, y) - b)
def polynomialKernel(x, y, d=2):
  return (1 + np.dot(x, y))**d

### 交差検定
# X: データ点の配列
# Y: データ点の属するクラスの配列
# k: 分割数
# kernel: カーネルトリック (デフォルトは内積)
def crossValidation(X, Y, k, kernel=np.dot):
  # answerを正解としてresultの正解率を計算
  def accuracyRate(result, answer):
    return 1.0 - sum(abs(result - answer)/2) / float(len(result))

  n = len(X)  # データ点の個数
  l = n / k   # k分割したデータ点の個数
  ac = 0.0    # 正解率の初期化

  # k-交差検定を行い正解率を計算
  for i in range(k):
    # l個の評価ベクトルとそのクラス
    testVectors = X[l*i:l*(i+1)]
    classForTestVectors = Y[l*i:l*(i+1)]
    # n-l個の訓練ベクトルとそのクラス
    learningVectors = np.vstack((X[:l*i], X[l*(i+1):]))
    classForlearningVectors = np.hstack((Y[:l*i], Y[l*(i+1):]))

    # 訓練ベクトルからサポートベクターを計算
    svm = SVM(learningVectors, classForlearningVectors, kernel)
    # 学習した識別関数で評価ベクトルを識別
    result = [svm.discriminate(t) for t in testVectors]
    # 評価結果の正解率を算出
    ac += accuracyRate(result, classForTestVectors)

  # 正解率の平均を返す
  return ac / k

if __name__ == '__main__':
  # コマンドライン引数処理
  if len(sys.argv)==1:
    kernel = np.dot
    D = np.loadtxt(sys.stdin)
  elif len(sys.argv)==3 and sys.argv[1]=='-k':
    if sys.argv[2]=='Gaussian':
      kernel = gaussianKernel
    elif sys.argv[2]=='Sigmoid':
      kernel = sigmoidKernel
    elif sys.argv[2]=='Polynomial':
      kernel = polynomialKernel
    else:
      print 'invalid kernel name (Gaussian/Sigmoid/Polynomial)'
      sys.exit(1)
    D = np.loadtxt(sys.stdin)
  else:
    print 'invalid arguments (must be the following style)'
    print '\'python '+__file__+' [-k KERNEL_NAME] < INPUT_FILE_PATH\''
    sys.exit(1)

  n = len(D)        # データ点の個数
  d = len(D[0])-1   # データ点の次元
  X = D[:, :d]      # データ点の配列
  Y = np.reshape(D[:, d:], n) #データ点の属するクラスの配列

  # サポートベクタ, 識別関数のプロット
  # svm = SVM(X, Y, kernel)
  # svm.plot()

  # 交差検定
  ar = crossValidation(X, Y, 10, kernel)
  # 正解率を出力
  print 'Accuracy Rate: ' + str(ar)

  